yarn-project/acir-simulator/src/avm/avm_context.ts

```diff
@@ -1,140 +1,63 @@
-    const forkedState = AvmJournal.branchParent(journal);
-    return new AvmContext(newExecutionEnvironment, forkedState);
+  public createNestedContractCallContext(address: AztecAddress, calldata: Fr[]): AvmContext {
+    const newExecutionEnvironment = this.environment.deriveEnvironmentForNestedCall(address, calldata);
+    const forkedWorldState = this.worldState.fork();
+    const machineState = AvmMachineState.fromState(this.machineState);
+    return new AvmContext(forkedWorldState, newExecutionEnvironment, machineState);
   }
 
   /**
-   * Prepare a new AVM context that will be ready for an external static call
-   * - It will fork the journal
-   * - It will set the correct execution Environment Variables for a call
-   *    - Alter both address and storageAddress
+   * Prepare a new AVM context that will be ready for an external/nested static call
+   * - Fork the world state journal
+   * - Derive a machine state from the current state
+   *   - E.g., gas metering is preserved but pc is reset
+   * - Derive an execution environment from the caller/parent
+   *   - Alter both address and storageAddress
    *
-   * @param address - The contract to call
-   * @param executionEnvironment - The current execution environment
-   * @param journal - The current journal
+   * @param address - The contract instance to initialize a context for
+   * @param calldata - Data/arguments for nested call
    * @returns new AvmContext instance
    */
-  public static prepExternalStaticCallContext(
-    address: AztecAddress,
-    calldata: Fr[],
-    executionEnvironment: AvmExecutionEnvironment,
-    journal: AvmJournal,
-  ): AvmContext {
-    const newExecutionEnvironment = executionEnvironment.newStaticCall(address, calldata);
-    const forkedState = AvmJournal.branchParent(journal);
-    return new AvmContext(newExecutionEnvironment, forkedState);
-  }
-
-  /**
-   * Merge the journal of this call with it's parent
-   * NOTE: this should never be called on a root context - only from within a nested call
-   */
-  public mergeJournalSuccess() {
-    this.journal.mergeSuccessWithParent();
-  }
-
-  /**
-   * Merge the journal of this call with it's parent
-   * For when the child call fails ( we still must track state accesses )
-   */
-  public mergeJournalFailure() {
-    this.journal.mergeFailureWithParent();
-  }
-}
-
-class NoBytecodeFoundInterpreterError extends AvmInterpreterError {
-  constructor(contractAddress: AztecAddress) {
-    super(`No bytecode found at: ${contractAddress}`);
-    this.name = 'NoBytecodeFoundInterpreterError';
+  public createNestedContractStaticCallContext(address: AztecAddress, calldata: Fr[]): AvmContext {
+    const newExecutionEnvironment = this.environment.deriveEnvironmentForNestedStaticCall(address, calldata);
+    const forkedWorldState = this.worldState.fork();
+    const machineState = AvmMachineState.fromState(this.machineState);
```

Comment:
Nested context needs to take in args for l1/l2/daGas as they aren't derived from the parent context but are actually arguments to the call/staticcall instructions.